src/devices/UFireIse/UFireIse.cs

```diff
@@ -0,0 +1,391 @@
+                WriteRegister((byte)Register.ISE_CALIBRATE_READHIGH_REGISTER, Convert.ToSingle(readHigh));
+            }
+        }
+
+        /// <summary>
+        /// Changes the default I2C address
+        /// </summary>
+        /// <param name="i2cAddress">The new I2C address</param>
+        public void SetI2CAddress(byte i2cAddress)
+        {
+            if (i2cAddress >= 1 && i2cAddress <= 127)
+            {
+                WriteRegister((byte)Register.ISE_SOLUTION_REGISTER, Convert.ToByte(i2cAddress));
+                SendCommand((byte)Command.ISE_I2C);
+            }
+            else
+            {
+                throw new ArgumentOutOfRangeException();
+            }
+        }
+
+        /// <summary>
+        /// Low level reads of EEPROM
+        /// </summary>
+        /// <param name="address">Address to read</param>
+        /// <returns>Low level data from EEPROM</returns>
+        public float ReadEeprom(byte address)
+        {
+            WriteRegister((byte)Register.ISE_SOLUTION_REGISTER, address);
+            SendCommand((byte)Command.ISE_MEMORY_READ);
+
+            return ReadRegister((byte)Register.ISE_BUFFER_REGISTER);
+        }
+
+        /// <summary>
+        /// Low level write of EEPROM
+        /// </summary>
+        /// <param name="address">Address</param>
+        /// <param name="data">Data to write to the address</param>
+        public void WriteEeprom(byte address, float data)
+        {
+            WriteRegister((byte)Register.ISE_SOLUTION_REGISTER, address);
+            WriteRegister((byte)Register.ISE_BUFFER_REGISTER, data);
+            SendCommand((byte)Command.ISE_MEMORY_WRITE);
+        }
+
+        /// <summary>
+        /// Get Firmware version
+        /// </summary>
+        /// <returns>The version of the firmware</returns>
+        public Version GetFirmwareVersion()
+        {
+            return new Version(ReadByte((byte)Register.ISE_FW_VERSION_REGISTER).ToString());
+
+        }
+
+        private void SendCommand(byte data)
+        {
+            Span<byte> bytes = stackalloc byte[2];
+            bytes[0] = (byte)Register.ISE_TASK_REGISTER;
+            bytes[1] = data;
+            _device.Write(bytes);
+            DelayHelper.DelayMilliseconds(IseCommunicationDelay, allowThreadYield: true);
+        }
+
+        private float ReadRegister(byte register)
+        {
+            ChangeRegister(register);
+
+            Span<byte> data = stackalloc byte[4];
+
+            _device.Read(data);
+
+            return Convert.ToSingle(RoundTotalDigits(BitConverter.ToSingle(data.ToArray(), 0), 7));
+        }
+
+        private void WriteRegister(byte register, float? data)
+        {
+            Span<byte> bytes = stackalloc byte[4]
+             {
+                0,
+                0,
+                0,
+                0
+             };
+
+            if (data != null)
+            {
+                bytes = BitConverter.GetBytes(RoundTotalDigits(data.Value));
```

Comment:
BitConverter.GetBytes will allocate array, can you use overload which takes Span and instead write to it?

Reply:
GetBytes do not have a Span overload

Reply:
@tannergooding do you perhaps know how to do this in non-allocating way?

Reply:
You want `TryWriteBytes`

Reply:
I now use TryWriteBytes